chain-tx-enclave/mls/src/keypackage.rs

```diff
@@ -0,0 +1,158 @@
+            .iter()
+            .filter_map(|ext| match ext {
+                Extension::Expiration(expire) => Some(*expire),
+                _ => None,
+            })
+            .next()
+            .ok_or(KeyPackageVerifyError::ExtensionNotFound("Expiration"))?;
+
+        if now > expire {
+            return Err(KeyPackageVerifyError::Expired(expire));
+        }
+
+        let verifier = EnclaveCertVerifier::new(ra_config)
+            .map_err(KeyPackageVerifyError::VerifierInitError)?;
+        verifier
+            .verify_cert(&self.credential)
+            .map_err(KeyPackageVerifyError::CertificateVerifyError)
+    }
+}
+
+/// Key package, only send `(payload, signature)` to other nodes.
+#[derive(Encode, Decode)]
+pub struct KeyPackage {
+    pub payload: KeyPackagePayload,
+    pub signature: Vec<u8>,
+    /// only have the secret key of our own key package
+    pub priv_key: Option<SecretKey>,
+}
+
+impl KeyPackage {
+    /// Create key package in enclave
+    #[cfg(feature = "edp")]
+    pub fn new(ra_ctx: EnclaveRaContext) -> Result<Self, EnclaveRaContextError> {
+        let expire = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("Time went backwards")
+            .as_secs()
+            + DEFAULT_LIFE_TIME;
+        let extensions = vec![
+            Extension::SupportedVersions(PROTOCOL_VERSION_MLS10),
+            Extension::Expiration(expire),
+        ];
+
+        let cert = ra_ctx.get_certificate()?;
+        let cert_private_key = cert.private_key.0.clone();
+        let priv_key = SecretKey::new(&mut rand::thread_rng());
+        let init_key = PublicKey::from_secret_key(&priv_key);
+        let payload = KeyPackagePayload {
+            version: PROTOCOL_VERSION_MLS10,
+            init_key,
+            credential: cert.certificate.0,
+            extensions,
+        };
+        let signature = payload.sign(&cert_private_key);
+        Ok(Self {
+            payload,
+            signature,
+            priv_key: Some(priv_key),
+        })
+    }
+
+    /// Verify key package and signature
+    pub fn verify(
+        &self,
+        ra_config: EnclaveCertVerifierConfig,
+        now: Timespec,
+    ) -> Result<(), KeyPackageVerifyError> {
+        let public_key = self.payload.verify(ra_config, now)?;
+        signature::ECDSA_P256_SHA256_ASN1
+            .verify(
+                public_key.as_slice().into(),
+                self.payload.encode().as_slice().into(),
+                self.signature.as_slice().into(),
+            )
+            .map_err(KeyPackageVerifyError::SignatureVerifyError)
+    }
+}
+
+/// Error type for key package verification.
+#[derive(thiserror::Error, Debug)]
+pub enum KeyPackageVerifyError {
+    #[error("signature verify error: {0}")]
+    SignatureVerifyError(ring::error::Unspecified),
+    #[error("required extension is not found: {0}")]
+    ExtensionNotFound(&'static str),
+    #[error("the key package is expired at: {0}")]
```

Comment:
key package expired at ?

Reply:
It'll output the expire timestamp of the key package if it's expired,  any suggestions on this?

Reply:
how about this?
 #[error("key package expired at: {0}")]

Reply:
Done